cpp/open3d/geometry/PointCloudSegmentation.cpp

```diff
@@ -29,6 +29,7 @@ class RandomSampler {
     explicit RandomSampler(const size_t total_size) : total_size_(total_size) {}
 
     std::vector<T> operator()(size_t sample_size) {
+        std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
I think it is sacrificing performance for reproducibility. The omp's performance gain will probably be negated by a mutex. 
If you want to keep this change, would you please show benchmark result (before and after) and show there is no degrade in speed? Otherwise, it may be better to use an option to toggle omp for reproducibility instead of introducing mutex.

Reply:
Hi @theNded , I have used the following code to benchmark the performace under macos m1 and linux intel i7-11700 :
```python
import numpy as np
import open3d as o3d
import time 

data = o3d.data.PLYPointCloud()
pcd = o3d.io.read_point_cloud(data.path)

num_iter = 1000
ts = []
for _ in range(100):
    o3d.utility.random.seed(1)
    pcd_down = pcd.voxel_down_sample(voxel_size=0.01)

    t0 = time.time()
    plane_model, inliers = pcd_down.segment_plane(distance_threshold=0.02,
                                                  ransac_n=3,
                                                  num_iterations=num_iter,
                                                  probability=1)
    duration = time.time() - t0
    ts.append(duration)

print(f'mean time {np.mean(np.array(ts))}')
```

The results (before and after) are very closed:
macos m1 : 0.030s - 0.032s
linux intel i7-11700: 0.090s - 0.093s

I think it is suitable to adopt this modification.

Reply:
Thank you for the clarification. I will merge this PR, but we still need to bear in mind potential problems. Please refer to this PR if you observe relevant issues.